### PR TITLE
Added orbit creation from Astropy SkyCoord in any reference frame

### PR DESCRIPTION
**Pull request by hrishikeshgoyal**
Addresses #434

### CODE REVIEW

File: src/poliastro/twobody/orbit.py [hunk 1]
```diff
@@ -228,6 +229,41 @@ def from_vectors(cls, attractor, r, v, epoch=J2000, plane=Planes.EARTH_EQUATOR):
...
+        ----------
+        attractor: Body
+            Main attractor
+        coord: astropy.coordinates.SkyCoord
+            Position and velocity vectors in any reference frame. Note that `SkyCoord` must have
+            a representation and its differential with respect to time.
+
+
+        """
+        if not isinstance(coord, SkyCoord):
```
Comment (Juanlu001) [inline src/poliastro/twobody/orbit.py]: I would remove the `isinstance` check. It's unpythonic (we should rely on "duck typing" instead) and also it's super slow.

File: src/poliastro/twobody/orbit.py [hunk 2]
```diff
@@ -228,6 +229,41 @@ def from_vectors(cls, attractor, r, v, epoch=J2000, plane=Planes.EARTH_EQUATOR):
...
+            raise ValueError("coord can only be an instance of SkyCoord class")
+
+        if "s" not in coord.cartesian.differentials:
```
Comment (Juanlu001) [inline src/poliastro/twobody/orbit.py]: Nice!

File: src/poliastro/twobody/orbit.py [hunk 3]
```diff
@@ -228,6 +229,41 @@ def from_vectors(cls, attractor, r, v, epoch=J2000, plane=Planes.EARTH_EQUATOR):
...
+            raise ValueError(
+                "SkyCoord instance passed as coord doesn't have a differential with respect to time"
+            )
+
+        # Get an inertial reference frame parallel to ICRS and centered at attractor
+        inertial_frame_at_body_centre = get_frame(attractor, Planes.EARTH_EQUATOR)
+
+        coord_in_irf = coord.transform_to(inertial_frame_at_body_centre)
```
Comment (Juanlu001) [inline src/poliastro/twobody/orbit.py]: Excellent! In this way, we don't have to add special checks or transformations in the propagation, and we can leverage the code we already have.

File: src/poliastro/twobody/orbit.py [hunk 4]
```diff
@@ -9,6 +9,7 @@
     CartesianDifferential,
     CartesianRepresentation,
     get_body_barycentric_posvel,
+    SkyCoord,
```
Comment (Juanlu001) [inline src/poliastro/twobody/orbit.py]: If we remove the `isinstance` check we won't need this import.

File: src/poliastro/tests/tests_twobody/test_orbit.py [hunk 5]
```diff
@@ -589,3 +589,37 @@ def test_pqw_returns_dimensionless():
...
     assert w.unit == u.one
+
+
+def test_orbit_from_skycoord_inputs():
+
+    pos = [30000, 0, 0] * u.km
+    cartrep = CartesianRepresentation(*pos)
+
+    # Method fails if coordinate is not passed as SkyCoord instance
+    with pytest.raises(ValueError) as excinfo:
```
Comment (Juanlu001) [inline src/poliastro/tests/tests_twobody/test_orbit.py]: And if we remove the `isinstance` check we don't need this part of the test either.

File: src/poliastro/tests/tests_twobody/test_orbit.py [hunk 6]
```diff
@@ -589,3 +589,37 @@ def test_pqw_returns_dimensionless():
...
+        ss = Orbit.from_skycoord(Earth, cartrep)
+    assert (
+        "ValueError: coord can only be an instance of SkyCoord class"
+        in excinfo.exconly()
+    )
+    # Method fails if SkyCoord instance doesn't contain a differential with respect to time
```
Comment (Juanlu001) [inline src/poliastro/tests/tests_twobody/test_orbit.py]: Better rename the test to `test_from_skycoord_fails_if_no_time_differential`

File: src/poliastro/tests/tests_twobody/test_orbit.py [hunk 7]
```diff
@@ -589,3 +589,37 @@ def test_pqw_returns_dimensionless():
...
+    with pytest.raises(ValueError) as excinfo:
+        ss = Orbit.from_skycoord(Earth, SkyCoord(cartrep))
+    assert (
+        "ValueError: SkyCoord instance passed as coord doesn't have a differential with respect to time"
+        in excinfo.exconly()
+    )
+
+
+def test_frame_of_orbit_from_skycoord():
```
Comment (Juanlu001) [inline src/poliastro/tests/tests_twobody/test_orbit.py]: Can we `pytest.mark.parametrize` this test to cover all the available bodies?

File: src/poliastro/tests/tests_twobody/test_orbit.py [hunk 8]
```diff
@@ -589,3 +589,37 @@ def test_pqw_returns_dimensionless():
...
+    pos = [30000, 0, 0] * u.km
+    cartrep = CartesianRepresentation(*pos)
+
+    vel = [2, 0, 0] * u.km / u.s
+    cartdiff = CartesianDifferential(*vel)
+
+    coord_with_diff = SkyCoord(cartrep.with_differentials(cartdiff), frame="icrs")
+    o = Orbit.from_skycoord(Earth, coord_with_diff)
+
+    assert isinstance(o.frame, GCRS)
```
Comment (Juanlu001) [inline src/poliastro/tests/tests_twobody/test_orbit.py]: Be careful with newlines

File: src/poliastro/tests/tests_twobody/test_orbit.py [hunk 9]
```diff
@@ -589,3 +589,37 @@ def test_pqw_returns_dimensionless():
...
+    with pytest.raises(ValueError) as excinfo:
+        ss = Orbit.from_skycoord(Earth, SkyCoord(cartrep))
+    assert (
+        "ValueError: SkyCoord instance passed as coord doesn't have a differential with respect to time"
+        in excinfo.exconly()
+    )
+
+
+def test_frame_of_orbit_from_skycoord():
+    pos = [30000, 0, 0] * u.km
```
Comment (Juanlu001) [inline src/poliastro/tests/tests_twobody/test_orbit.py]: We should also check that `orbit.r == pos` and `orbit.v == vel` for completeness.

File: src/poliastro/twobody/orbit.py (hunk 1 above)
Comment (hrishikeshgoyal) [reply src/poliastro/twobody/orbit.py]: Sure, removed

File: src/poliastro/twobody/orbit.py (hunk 3 above)
Comment (hrishikeshgoyal) [reply src/poliastro/twobody/orbit.py]: Thanks, yes that was the motive. It shouldn't force us to implement propagation in the user reference frame. At the same time, later we should be able to implement it if we want.

File: src/poliastro/tests/tests_twobody/test_orbit.py (hunk 5 above)
Comment (hrishikeshgoyal) [reply src/poliastro/tests/tests_twobody/test_orbit.py]: Removed.

File: src/poliastro/tests/tests_twobody/test_orbit.py (hunk 6 above)
Comment (hrishikeshgoyal) [reply src/poliastro/tests/tests_twobody/test_orbit.py]: Done

File: src/poliastro/tests/tests_twobody/test_orbit.py (hunk 9 above)
Comment (hrishikeshgoyal) [reply src/poliastro/tests/tests_twobody/test_orbit.py]: Added this check

File: src/poliastro/tests/tests_twobody/test_orbit.py (hunk 7 above)
Comment (hrishikeshgoyal) [reply src/poliastro/tests/tests_twobody/test_orbit.py]: Done

File: src/poliastro/tests/tests_twobody/test_orbit.py (hunk 8 above)
Comment (hrishikeshgoyal) [reply src/poliastro/tests/tests_twobody/test_orbit.py]: We shouldn't have a newline at EOF right?

File: src/poliastro/tests/tests_twobody/test_orbit.py [hunk 10]
```diff
@@ -589,3 +595,68 @@ def test_pqw_returns_dimensionless():
...
+
+    coord = frame(cartrep, obstime=J2000)
+    o = Orbit.from_coord(attractor, coord)
+
+    inertial_frame_at_body_centre = get_frame(attractor, Planes.EARTH_EQUATOR)
+
+    coord_transformed_to_irf = coord.transform_to(inertial_frame_at_body_centre)
+
+    pos_transformed_to_irf = coord_transformed_to_irf.cartesian.xyz
+    vol__transformed_to_irf = coord_transformed_to_irf.cartesian.differentials[
```
Comment (Juanlu001) [inline src/poliastro/tests/tests_twobody/test_orbit.py]:
```suggestion
    vel_transformed_to_irf = coord_transformed_to_irf.cartesian.differentials[
```

File: src/poliastro/twobody/orbit.py [hunk 11]
```diff
@@ -228,6 +228,38 @@ def from_vectors(cls, attractor, r, v, epoch=J2000, plane=Planes.EARTH_EQUATOR):
         ss = RVState(attractor, r, v)
         return cls(ss, epoch, plane)
 
+    @classmethod
+    def from_coord(cls, attractor, coord):
```
Comment (Juanlu001) [inline src/poliastro/twobody/orbit.py]:
```suggestion
    def from_coords(cls, attractor, coord):
```

File: src/poliastro/twobody/orbit.py [hunk 12]
```diff
@@ -228,6 +228,38 @@ def from_vectors(cls, attractor, r, v, epoch=J2000, plane=Planes.EARTH_EQUATOR):
...
+        This method accepts `SkyCoord` in any reference frame unlike `Orbit.from_vector`
+        which can accept inputs in only inertial reference frame centred at attractor.
+        Also note that the frame information is lost after creation of the orbit and
+        only the inertial reference frame at body centre will be used for all purposes.
+        Parameters
+        ----------
+        attractor: Body
+            Main attractor
+        coord: astropy.coordinates.SkyCoord
+            Position and velocity vectors in any reference frame. Note that `SkyCoord` must have
```
Comment (Juanlu001) [inline src/poliastro/twobody/orbit.py]: There is some error in this docstring: `sphinx.errors.SphinxWarning: /home/circleci/repo/venv/lib/python3.6/site-packages/poliastro/twobody/orbit.py:docstring of poliastro.twobody.orbit.Orbit.from_coord:9:Unexpected indentation.`

https://circleci.com/gh/poliastro/poliastro/1936

Please do

```
sphinx-build -vW -b html docs/source docs-html
```

locally to make sure it's fixed.

**Comment by Juanlu001** [comment]
First problem:

```python
>>> coords = GCRS(CartesianRepresentation([[1, 0, 0], [.9, .1, 0]] * u.km, differentials=CartesianDifferential([[0, 1, 0], [-.1, .9, 0]] * u.km / u.s, xyz_axis=1), xyz_axis=1), representation_type=CartesianRepresentation)
>>> Orbit.from_coords(Earth, coords)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/juanlu/Development/poliastro/poliastro-library/src/poliastro/twobody/orbit.py", line 258, in from_coords
    coord_in_irf = coord.transform_to(inertial_frame_at_body_centre)
  File "/home/juanlu/.miniconda36/envs/poliastro37/lib/python3.7/site-packages/astropy/coordinates/baseframe.py", line 1163, in transform_to
    raise NotImplementedError('You cannot transform a frame that has '
NotImplementedError: You cannot transform a frame that has velocities to another frame at a different obstime. If you think this should (or should not) be possible, please comment at https://github.com/astropy/astropy/issues/6280
```

Second problem:

```python
>>> coords = GCRS(CartesianRepresentation([[1, 0, 0], [.9, .1, 0]] * u.km, differentials=CartesianDifferential([[0, 1, 0], [-.1, .9, 0]] * u.km / u.s, xyz_axis=1), xyz_axis=1), representation_type=CartesianRepresentation, obstime=J2000)
>>> Orbit.from_coords(Earth, coords)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/juanlu/Development/poliastro/poliastro-library/src/poliastro/twobody/orbit.py", line 662, in __repr__
    return self.__str__()
  File "/home/juanlu/Development/poliastro/poliastro-library/src/poliastro/twobody/orbit.py", line 636, in __str__
    if self.a > 1e7 * u.km:
  File "/home/juanlu/Development/poliastro/poliastro-library/src/poliastro/twobody/orbit.py", line 105, in a
    return self.p / (1 - self.ecc ** 2)
  File "/home/juanlu/Development/poliastro/poliastro-library/src/poliastro/twobody/orbit.py", line 110, in p
    return self._state.to_classical().p
  File "/home/juanlu/Development/poliastro/poliastro-library/src/poliastro/twobody/_states.py", line 181, in to_classical
    self.v.to(u.km / u.s).value,
  File "/home/juanlu/.miniconda36/envs/poliastro37/lib/python3.7/site-packages/numba/dispatcher.py", line 348, in _compile_for_args
    error_rewrite(e, 'typing')
  File "/home/juanlu/.miniconda36/envs/poliastro37/lib/python3.7/site-packages/numba/dispatcher.py", line 315, in error_rewrite
    reraise(type(e), e, None)
  File "/home/juanlu/.miniconda36/envs/poliastro37/lib/python3.7/site-packages/numba/six.py", line 658, in reraise
    raise value.with_traceback(tb)
numba.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Failed in nopython mode pipeline (step: nopython frontend)
Invalid use of Function(<built-in function array>) with argument(s) of type(s): (tuple(array(float64, 1d, C) x 3))
 * parameterized
In definition 0:
    TypingError: array(float64, 1d, C) not allowed in a homogeneous sequence
    raised from /home/juanlu/.miniconda36/envs/poliastro37/lib/python3.7/site-packages/numba/typing/npydecl.py:463
In definition 1:
    TypingError: array(float64, 1d, C) not allowed in a homogeneous sequence
    raised from /home/juanlu/.miniconda36/envs/poliastro37/lib/python3.7/site-packages/numba/typing/npydecl.py:463
This error is usually caused by passing an argument of a type that is unsupported by the named function.
[1] During: resolving callee type: Function(<built-in function array>)
[2] During: typing of call at /home/juanlu/Development/poliastro/poliastro-library/src/poliastro/core/util.py (166)


File "src/poliastro/core/util.py", line 166:
def cross(a, b):
    <source elided>
            a[2] * b[0] - a[0] * b[2],
            a[0] * b[1] - a[1] * b[0],
            ^

[1] During: resolving callee type: type(CPUDispatcher(<function cross at 0x7f57c3cf96a8>))
[2] During: typing of call at /home/juanlu/Development/poliastro/poliastro-library/src/poliastro/core/elements.py (274)


File "src/poliastro/core/elements.py", line 274:
def rv2coe(k, r, v, tol=1e-8):
    <source elided>

    h = cross(r, v)
    ^

This is not usually a problem with Numba itself but instead often caused by
the use of unsupported features or an issue in resolving types.

To see Python/NumPy features supported by the latest release of Numba visit:
http://numba.pydata.org/numba-doc/dev/reference/pysupported.html
and
http://numba.pydata.org/numba-doc/dev/reference/numpysupported.html

For more information about typing errors and how to debug them visit:
http://numba.pydata.org/numba-doc/latest/user/troubleshoot.html#my-code-doesn-t-compile

If you think your code should work with Numba, please report the error message
and traceback, along with a minimal reproducer at:
https://github.com/numba/numba/issues/new

```

Before fixing them, please turn these two cases (or similar) into unit tests :wink:

**Comment by hrishikeshgoyal** [comment]
Just to confirm, do we need to support orbit creation from an array of position and velocity vectors?

File: src/poliastro/tests/tests_twobody/test_orbit.py (hunk 8 above)
Comment (Juanlu001) [reply src/poliastro/tests/tests_twobody/test_orbit.py]: Yes we should :)

Comment (Juanlu001) [reply src/poliastro/tests/tests_twobody/test_orbit.py]: now it's ok

**Comment by Juanlu001** [comment]
> Just to confirm, do we need to support orbit creation from an array of position and velocity vectors?

No, but we should fail gracefully :)

**Comment by hrishikeshgoyal** [comment]
Alright, that makes sense :)

**Comment by Juanlu001** [comment]
> In that case should we transform the coords to inertial frame at same obstime?

Yes! And the `epoch` of the final orbit should be that `obstime`.

With that, this PR would be complete.

File: src/poliastro/twobody/orbit.py [hunk 13]
```diff
@@ -271,7 +273,7 @@ def from_coords(cls, attractor, coord):
         pos = coord_in_irf.cartesian.xyz
         vel = coord_in_irf.cartesian.differentials["s"].d_xyz
 
-        return cls.from_vectors(attractor, pos, vel)
+        return cls.from_vectors(attractor, pos, vel, epoch=coord.obstime)
```
Comment (Juanlu001) [inline src/poliastro/twobody/orbit.py]: Minor nitpick: we are not testing that `orbit.epoch == coords.obstime` :)

Comment (hrishikeshgoyal) [reply src/poliastro/twobody/orbit.py]: yes. I missed that :(

**Pull request merged**